// @ephox/katamari/lib/test/ts/module/ephox/katamari/test/arb/ArbDataTypes.js
import fc from 'fast-check';
import * as Fun from 'ephox/katamari/api/Fun';
import { Future } from 'ephox/katamari/api/Future';
import { Optional } from 'ephox/katamari/api/Optional';
import { Result } from 'ephox/katamari/api/Result';
export const arbResultError = (arbE) => arbE.map(Result.error);
export const arbResultValue = (arbA) => arbA.map(Result.value);
export const arbResult = (arbA, arbE) => fc.oneof(arbResultError(arbE), arbResultValue(arbA));
export const arbOptionalNone = () => fc.constant(Optional.none());
export const arbOptionalSome = (at) => at.map(Optional.some);
export const arbOptional = (at) => fc.oneof(arbOptionalNone(), arbOptionalSome(at));
export const arbNegativeInteger = () => fc.integer(Number.MIN_SAFE_INTEGER, -1);
export const arbFutureNow = (arbA) => arbA.map(Future.pure);
export const arbFutureSoon = (arbA) => arbA.map((a) => Future.nu((cb) => {
    setTimeout(() => {
        cb(a);
    }, 5);
}));
export const arbFutureNever = () => fc.constant(Future.nu(Fun.noop));
export const arbFutureNowOrSoon = (arbA) => fc.oneof(arbFutureNow(arbA), arbFutureSoon(arbA));
