// @ephox/agar/lib/main/ts/ephox/agar/api/Log.js
import { Arr } from '@ephox/katamari';
import { Chain } from './Chain';
import { sequence } from './GeneralSteps';
import { Step } from './Step';
const generateLogMsg = (testId, description) => 
// AP-147 Format: 'TestCase-<plugin name>-<test case ID / TBA:> <description of the test>'
`TestCase-${testId}: ${description}`;
const step = (testId, description, f) => Step.label(generateLogMsg(testId, description), f);
const steps = (testId, description, fs) => Arr.map(fs, (f, i) => step(testId, description + ' (' + i + ')', f));
const stepsAsStep = (testId, description, fs) => sequence(steps(testId, description, fs));
const chain = (testId, description, c) => Chain.label(generateLogMsg(testId, description), c);
const chains = (testId, description, cs) => Arr.map(cs, (c, i) => chain(testId, description + ' (' + i + ')', c));
const chainsAsChain = (testId, description, cs) => Chain.fromChains(chains(testId, description, cs));
const chainsAsStep = (testId, description, cs) => Chain.asStep({}, chains(testId, description, cs));
export { step, steps, stepsAsStep, chain, chains, chainsAsChain, chainsAsStep };
