// @ephox/agar/lib/main/ts/ephox/agar/api/Touch.js
import { Focus } from '@ephox/sugar';
import * as Touches from '../touch/Touches';
import { Chain } from './Chain';
import { Step } from './Step';
import * as UiFinder from './UiFinder';
const touchStart = Touches.touchstart;
const touchStartAt = (element, dx, dy) => Touches.touchstartAt(dx, dy)(element);
const touchEnd = Touches.touchend;
const touchEndAt = (element, dx, dy) => Touches.touchendAt(dx, dy)(element);
const touchMove = Touches.touchmove;
const touchMoveTo = (element, dx, dy) => Touches.touchmoveTo(dx, dy)(element);
const cTrigger = (selector, action) => Chain.async((container, next, die) => {
    UiFinder.findIn(container, selector).fold(() => die('Could not find element: ' + selector), (ele) => {
        action(ele);
        next(container);
    });
});
const sTriggerWith = (container, selector, action) => Chain.asStep(container, [cTrigger(selector, action)]);
const trueTap = (elem) => {
    // The closest event queue to a true tap event
    Focus.focus(elem);
    Touches.touchstart(elem);
    Touches.touchend(elem);
};
const tap = (elem) => {
    Touches.touchstart(elem);
    Touches.touchend(elem);
};
const sTap = (element) => Step.sync(() => tap(element));
const sTrueTapOn = (container, selector) => sTriggerWith(container, selector, trueTap);
const sTapOn = (container, selector) => sTriggerWith(container, selector, tap);
const cTapOn = (selector) => cTrigger(selector, tap);
const cTouchStartAt = (dx, dy) => Chain.op(Touches.touchstartAt(dx, dy));
const cTouchEndAt = (dx, dy) => Chain.op(Touches.touchendAt(dx, dy));
const cTouchMoveTo = (dx, dy) => Chain.op(Touches.touchmoveTo(dx, dy));
const point = Touches.point;
const cTrueTap = Chain.op(trueTap);
const cTap = Chain.op(tap);
const cTouchMove = Chain.op(Touches.touchmove);
const cTouchStart = Chain.op(Touches.touchstart);
const cTouchEnd = Chain.op(Touches.touchend);
export { point, tap, trueTap, touchStart, touchStartAt, touchEnd, touchEndAt, touchMove, touchMoveTo, sTap, sTapOn, sTrueTapOn, cTap, cTapOn, cTrueTap, cTouchStart, cTouchStartAt, cTouchMove, cTouchMoveTo, cTouchEnd, cTouchEndAt };
